### PR TITLE
bpo-39406: Implement os.putenv() with setenv() if available

### PR DESCRIPTION
**Pull request by vstinner**
If setenv() C function is available, os.putenv() is now implemented
with setenv() instead of putenv(), so Python doesn't have to handle
the environment variable memory.

If setenv() is available, posix_putenv_garbage state is not needed.

https://bugs.python.org/issue39406

### CODE REVIEW

File: configure [hunk 1]
```diff
@@ -896,7 +895,6 @@ datadir='${datarootdir}'
 sysconfdir='${prefix}/etc'
 sharedstatedir='${prefix}/com'
 localstatedir='${prefix}/var'
-runstatedir='${localstatedir}/run'
```
Comment (serhiy-storchaka) [inline configure:899]: Unrelated changes (you use different version of autotools).

File: Modules/posixmodule.c [hunk 2]
```diff
@@ -819,9 +819,15 @@ dir_fd_converter(PyObject *o, void *p)
     }
 }
 
+#if defined(HAVE_PUTENV) && (defined(MS_WINDOWS) || !defined(HAVE_SETENV))
```
Comment (serhiy-storchaka) [inline Modules/posixmodule.c]: Names like `HAVE_SOMETHING` usually either defined or undefined. But names like `USE_SOMETHING` usually are defined as 1 or 0. It is not a strict rule.
```c
#if defined(HAVE_PUTENV) && (defined(MS_WINDOWS) || !defined(HAVE_SETENV))
#  define USE_PUTENV_GARBAGE 1
#else
#  define USE_PUTENV_GARBAGE 0
#endif
```

File: Modules/posixmodule.c [hunk 3]
```diff
@@ -10062,8 +10072,10 @@ posix_putenv_garbage_setitem(PyObject *name, PyObject *value)
     else
         Py_DECREF(value);
 }
+#endif
```
Comment (serhiy-storchaka) [inline Modules/posixmodule.c]:
```suggestion
#endif  /* USE_PUTENV_GARBAGE */
```

File: Modules/posixmodule.c [hunk 4]
```diff
@@ -10200,6 +10223,8 @@ os_unsetenv_impl(PyObject *module, PyObject *name)
         }
         PyErr_Clear();
     }
+#endif
```
Comment (serhiy-storchaka) [inline Modules/posixmodule.c]:
```suggestion
#endif  /* USE_PUTENV_GARBAGE */
```

**Review by serhiy-storchaka — APPROVED** [review]

File: configure (hunk 1 above)
Comment (vstinner) [reply configure:899]: I have no idea how to prevent such unrelated change. It was discussed multiple times, no solution was found yet.

File: Modules/posixmodule.c (hunk 2 above)
Comment (vstinner) [reply Modules/posixmodule.c]: I renamed the variable to PY_PUTENV_DICT.

**Comment by vstinner** [comment]
I updated the PR to use SetEnvironmentVariable() on Windows.

@serhiy-storchaka: Would you mind to review the updated PR?

**Comment by vstinner** [comment]
On Windows, I removed the explicit check on _MAX_ENV:

```
    if (size > _MAX_ENV) {
        PyErr_Format(PyExc_ValueError,
                     "the environment variable is longer than %u characters",
                     _MAX_ENV);
        goto error;
    }
```

On my Windows 10 VM, _MAX_ENV is equal to 32767, but SetEnvironmentVariableW() seems to accept way longer values!

```
>>> value='x' * (32767); os.environ['long'] = value; os.environ['long'] == value, len(value)    
(True, 32767)
>>> value='x' * (32767+10); os.environ['long'] = value; os.environ['long'] == value, len(value) 
(True, 32777)
>>> value='abcd' * (32767); os.environ['long'] = value; os.environ['long'] == value, len(value) 
(True, 131068)
```

But I got ERROR_NO_SYSTEM_RESOURCES (error 1450: "Insufficient system resources exist to complete the requested service") for a value of 1 GB:

```
>>> import os
>>> value='x'*(2**30)
>>> os.putenv('long', value)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [WinError 1450] Ressources système insuffisantes pour terminer le service demandé
```

The limit seems to be the memory rather than an hardcoded limit. A "shorter" value (64 MB) is fine:

```
>>> import os
>>> value='x'*(2**26)
>>> os.putenv('long', value)
```

File: Modules/posixmodule.c [hunk 5]
```diff
@@ -819,9 +819,18 @@ dir_fd_converter(PyObject *o, void *p)
...
+
 typedef struct {
     PyObject *billion;
-    PyObject *posix_putenv_garbage;
+#ifdef PY_PUTENV_DICT
+    /* putenv() requires to manage the memory manually: use a Python dict
+       object mapping name (bytes) to env (bytes) where env is a bytes string
+       like "name=value\0". */
+    PyObject *posix_putenv_dict;
```
Comment (serhiy-storchaka) [inline Modules/posixmodule.c:832]: Why rename it? This increases the size of the diff.

File: Modules/posixmodule.c [hunk 6]
```diff
@@ -10090,38 +10101,29 @@ os_putenv_impl(PyObject *module, PyObject *name, PyObject *value)
...
 
-    posix_putenv_garbage_setitem(name, unicode);
     Py_RETURN_NONE;
-
-error:
-    Py_DECREF(unicode);
-    return NULL;
 }
-#else /* MS_WINDOWS */
+#elif defined(HAVE_PUTENV)
```
Comment (serhiy-storchaka) [inline Modules/posixmodule.c:10126]:
```suggestion
#elif defined(HAVE_SETENV) || defined(HAVE_PUTENV)
```

File: Modules/posixmodule.c [hunk 7]
```diff
@@ -10136,31 +10138,38 @@ static PyObject *
...
+#ifdef PY_PUTENV_DICT
+    posix_putenv_dict_setitem(name, bytes);
+#endif
+#endif  /* !HAVE_SETENV */
+
     Py_RETURN_NONE;
 }
-#endif /* MS_WINDOWS */
-#endif /* HAVE_PUTENV */
+#endif /* HAVE_PUTENV  */
```
Comment (serhiy-storchaka) [inline Modules/posixmodule.c:10172]:
```suggestion
#endif  /* HAVE_SETENV || HAVE_PUTENV */
```

File: configure (hunk 1 above)
Comment (serhiy-storchaka) [reply configure:899]: The simplest way is to remove unrelated changes manually.

**Comment by vstinner** [comment]
Oh, a test fail on Windows:

```

2020-01-21T10:50:49.1283745Z ======================================================================
2020-01-21T10:50:49.1283887Z ERROR: test_unset_error (test.test_os.EnvironTests)
2020-01-21T10:50:49.1284029Z ----------------------------------------------------------------------
2020-01-21T10:50:49.1284166Z Traceback (most recent call last):
2020-01-21T10:50:49.1284296Z   File "D:\a\cpython\cpython\lib\test\support\__init__.py", line 695, in wrapper
2020-01-21T10:50:49.1284429Z     return func(*args, **kw)
2020-01-21T10:50:49.1284504Z   File "D:\a\cpython\cpython\lib\test\test_os.py", line 962, in test_unset_error
2020-01-21T10:50:49.1284628Z     self.assertRaises(ValueError, os.environ.__delitem__, key)
2020-01-21T10:50:49.1284764Z   File "D:\a\cpython\cpython\lib\unittest\case.py", line 799, in assertRaises
2020-01-21T10:50:49.1284895Z     return context.handle('assertRaises', args, kwargs)
2020-01-21T10:50:49.1285989Z   File "D:\a\cpython\cpython\lib\unittest\case.py", line 202, in handle
2020-01-21T10:50:49.1286157Z     callable_obj(*args, **kwargs)
2020-01-21T10:50:49.1286268Z   File "D:\a\cpython\cpython\lib\os.py", line 691, in __delitem__
2020-01-21T10:50:49.1286377Z     raise KeyError(key) from None
2020-01-21T10:50:49.1295463Z KeyError: 
```

**Comment by vstinner** [comment]
I made other changes in the meanwhile which introduced conflicts. So I wrote a new PR instead: PR #18128.
